### PR TITLE
Provison Azure Service

### PR DESCRIPTION
**Pull request by dilin-MS**
1. Bug fix: Enable return to last step when discovering device
2. Scaffold common files when load an iot project
3. Add "Provison Azure Service"
4. Create DevKit project with Azure Function
5. Archieve Dockerfiles on ACR

### CODE REVIEW

File: src/FileUtility.ts [hunk 1]
```diff
@@ -0,0 +1,114 @@
+'use strict';
+
+import * as fs from 'fs-plus';
+import * as path from 'path';
+import * as sdk from 'vscode-iot-device-cube-sdk';
+
+export class FileUtility {
+
+  // File-related functions to help create files on workspace path
+  static async existsInWorkspace(localPath: string): Promise<boolean> {
```
Comment (lirenhe) [inline src/FileUtility.ts]: How is this path related to workspace?

Comment (lirenhe) [reply src/FileUtility.ts]: I think the api signature should be changed to
fileExists(path, location).    If location is local, use the API from fs-plus, else use the API from cube.

File: src/Models/AZ3166Device.ts [hunk 2]
```diff
@@ -206,8 +206,9 @@ export class AZ3166Device extends ArduinoDeviceBase {
       throw new Error('Unable to find the board in the config file.');
     }
 
-    // await ScaffoldGenerator.scaffolIoTProjectdFiles(this.projectFolder, this.vscodeFolderPath, 
-    //   this.boardFolderPath, this.devcontainerFolderPath, this.board.id);
+    const scaffoldGenerator = new ScaffoldGenerator();
+    await scaffoldGenerator.scaffolIoTProjectdFilesInWorkspace(this.projectFolder, this.vscodeFolderPath,
```
Comment (lirenhe) [inline src/Models/AZ3166Device.ts]: scaffolIoTProjectdFilesInWorkspace  typo

what is the return value of this function?

File: src/Models/AzureComponentConfig.ts [hunk 3]
```diff
@@ -40,38 +40,60 @@ export class AzureConfigFileHandler {
...
   constructor(projectRoot: string) {
     this.projectRootPath = projectRoot;
     this.configFilePath = path.join(
         this.projectRootPath, AzureComponentsStorage.folderName,
         AzureComponentsStorage.fileName);
   }
 
-  createIfNotExists() {
+
+  async createIfNotExistsInWorkspace() {
```
Comment (lirenhe) [inline src/Models/AzureComponentConfig.ts]: combine those 2 API, use a variable to indicate whether this is local/remote in constructor.

File: src/Models/AzureComponentConfig.ts [hunk 4]
```diff
@@ -107,11 +129,10 @@ export class AzureConfigFileHandler {
     }
   }
 
-  getComponentIndexById(id: string) {
+  async getComponentIndexById(id: string) {
     try {
-      const azureConfigs =
-          JSON.parse(fs.readFileSync(this.configFilePath, 'utf8')) as
-          AzureConfigs;
+      const azureConfigContent = await FileUtility.readFileInLocal(this.configFilePath, 'utf8');
```
Comment (lirenhe) [inline src/Models/AzureComponentConfig.ts]: no idea why use async here. any benefit?

File: src/Models/AzureComponentConfig.ts [hunk 5]
```diff
@@ -120,11 +141,10 @@ export class AzureConfigFileHandler {
     }
   }
 
-  getComponentById(id: string) {
+  async getComponentById(id: string) {
```
Comment (lirenhe) [inline src/Models/AzureComponentConfig.ts]: same above. No need to be async

File: resources/Arduino/devkit/template.json [hunk 6]
```diff
@@ -12,6 +12,19 @@
...
+            "detail": "Create project that connects to Azure IoT Hub and processes device data further in Azure Functions.",
+            "label": "With Azure Functions",
+            "type": "AzureFunctions",
+            "templateFilesInfo": [
+                {
+                    "fileName": "simple_iothub.ino",
+                    "sourcePath" : ".",
+                    "targetPath" : "./device"
+                }
+            ]
```
Comment (lirenhe) [inline resources/Arduino/devkit/template.json]: For the support of Azure Functions, there are some additional tool to be installed on target machine.
https://marketplace.visualstudio.com/items?itemName=ms-azuretools.vscode-azurefunctions
Does your docker file contain all those information?

Comment (dilin-MS) [reply resources/Arduino/devkit/template.json]: The current version of iot workbench will suggest install the azure-function extension when user choose "With Azure Functions" devkit project. This method encounters one issue that currently vscode-insider does not provide accessible link like` vscode:extension/ms-azuretools.vscode-azurefunctions`.

As you suggested, we can directly choose a docker image with azure function extension pre-installed for users if they choose to create a "With Azure Functions" devkit project.
I will fix the issue next working day.

File: src/FileUtility.ts (hunk 1 above)
Comment (dilin-MS) [reply src/FileUtility.ts]: * **Workspace path**  looks like: `/workspace/myproject/devkit/sample.ino`
* **Local path** looks like: `C:/users/username/Documents/IoTWorkbenchProjects/projects/myproject/devkit/sample.ino`

I use `async` to align two sets of API(for local path and workspace path) which will save much duplicate code. And the local file-related functions is said will be refactored to be `async`. Did I misunderstand that?

Also, your suggestion to pass the path as an argument may be a better implementation. I will check whether this method can work well in the specific scenario.

File: src/Models/AzureComponentConfig.ts (hunk 4 above)
Comment (dilin-MS) [reply src/Models/AzureComponentConfig.ts]: The reason is that API we used from fs-plus is **sync**, but API from cube is **async**. If we combine the two API into one, we have to use **async** since we are not certain which API will be used.

File: resources/Arduino/devkit/template.json (hunk 6 above)
Comment (dilin-MS) [reply resources/Arduino/devkit/template.json]: Fixed.

File: resources/Linux/raspberrypi/IoTHub/devcontainer.json [hunk 7]
```diff
@@ -0,0 +1,10 @@
+{
+  "image": "rpi_app",
+  "name": "rpi_app_container",
+  "dockerFile": "Dockerfile",
+  "extensions": [
+    "ms-vscode.cpptools",
+    "twxs.cmake",
```
Comment (lirenhe) [inline resources/Linux/raspberrypi/IoTHub/devcontainer.json]: Why is this extension needed?

Comment (dilin-MS) [reply resources/Linux/raspberrypi/IoTHub/devcontainer.json]: Because we have CMakeLists.txt as one of our sample files. This extension is for intelliSense of CMake file.

File: src/FileUtility.ts [hunk 8]
```diff
@@ -0,0 +1,107 @@
+'use strict';
+
+import * as fs from 'fs-plus';
+import * as path from 'path';
+import * as sdk from 'vscode-iot-device-cube-sdk';
+import {ScaffoldType} from './constants';
+
+export class FileUtility {
+
+  static async exists(type: ScaffoldType, path: string): Promise<boolean> {
```
Comment (lirenhe) [inline src/FileUtility.ts]: better call it
  static async fileExists(environmentType: EnvironmentType, path: string): Promise<boolean> {

Comment (dilin-MS) [reply src/FileUtility.ts]: This functions can judge not only **file** but also **directory**. So we use `exists` as the function name.
We can change the `type` argument to `environmentType`.

File: src/FileUtility.ts [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    if (type === ScaffoldType.local) {
+      return await sdk.FileSystem.exists(path);
+    } else {
+      return new Promise((resolve: (exist: boolean) => void) => {
```
Comment (lirenhe) [inline src/FileUtility.ts]: I think you could still use fs.statsync in this case to avoid the callback.
You also need to check the return value to detect whether this is a file or a dir.

Comment (lirenhe) [reply src/FileUtility.ts]: I found there is a directoryExists api below, what does this API plan to detect?

File: src/FileUtility.ts (hunk 8 above)
Comment (dilin-MS) [reply src/FileUtility.ts]: But actually, the type is not indicating the environment whether is local or in workspace, since scenario can be that when we are in remote side and want to create a new project. In this situation we will pass a `local` type as argument in order to use the cube api to scaffold files in local filesystem.
i.e. the `type` argument is more about where to scaffold the new files, in local filesystem or in workspace.
Will it be better if we use the `scaffoldType` as the argument name?

File: src/constants.ts [hunk 10]
```diff
@@ -57,6 +58,11 @@ export enum OperationType {
   upload = 'Upload device code'
 }
 
+export enum ScaffoldType {
```
Comment (lirenhe) [inline src/constants.ts]: use EnvironmentType with value: local, container, wsl, remotemachine

File: src/Models/AzureComponentConfig.ts [hunk 11]
```diff
@@ -47,31 +47,29 @@ export class AzureConfigFileHandler {
         AzureComponentsStorage.fileName);
   }
 
-  createIfNotExists() {
+  async createIfNotExists(type: ScaffoldType) {
     const azureConfigs: AzureConfigs = {componentConfigs: []};
     const azureConfigFolderPath =
         path.join(this.projectRootPath, AzureComponentsStorage.folderName);
-    if (!fs.existsSync(azureConfigFolderPath)) {
+    if (!await FileUtility.exists(type, azureConfigFolderPath)) {
```
Comment (lirenhe) [inline src/Models/AzureComponentConfig.ts]: use dir exists?

File: src/Models/AzureComponentConfig.ts [hunk 12]
```diff
@@ -107,11 +105,10 @@ export class AzureConfigFileHandler {
     }
   }
 
-  getComponentIndexById(id: string) {
+  async getComponentIndexById(id: string) {
     try {
-      const azureConfigs =
-          JSON.parse(fs.readFileSync(this.configFilePath, 'utf8')) as
-          AzureConfigs;
+      const azureConfigContent = await FileUtility.readFile(ScaffoldType.local, this.configFilePath, 'utf8');
```
Comment (lirenhe) [inline src/Models/AzureComponentConfig.ts]: why using scaffoldtype.local here and workspace in async getSortedComponents

File: src/FileUtility.ts (hunk 9 above)
Comment (dilin-MS) [reply src/FileUtility.ts]: The `exists` function can be removed. Fixed.

File: src/Models/AzureComponentConfig.ts (hunk 11 above)
Comment (dilin-MS) [reply src/Models/AzureComponentConfig.ts]: Remove the `exists` api and fix this issue.

File: src/FileUtility.ts [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+  static async directoryExists(type: ScaffoldType, dirPath: string): Promise<boolean> {
+    if (type === ScaffoldType.local) {
+      if (!await sdk.FileSystem.exists(dirPath)) {
+        return false;
+      }
+      const isDirectory = await sdk.FileSystem.isDirectory(dirPath);
+      return isDirectory;
+    } else {
+      try {
+        return fs.statSync(dirPath).isDirectory();
```
Comment (Sneezry) [inline src/FileUtility.ts]: Do not use sync method

File: src/FileUtility.ts [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+        return false;
+      }
+    }
+  }
+
+  static async mkdir(type: ScaffoldType, dirPath: string): Promise<void> {
+    if (type === ScaffoldType.local) {
+      await sdk.FileSystem.mkDir(dirPath);
+    } else {
+      fs.mkdirSync(dirPath);
```
Comment (Sneezry) [inline src/FileUtility.ts]: Do not use sync method

File: src/FileUtility.ts [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+    }
+  }
+
+  static async mkdirRecursively(type: ScaffoldType, dirPath: string): Promise<void> {
+    if (await FileUtility.directoryExists(type, dirPath)) {
+      return;
+    }
+    const dirname = path.dirname(dirPath);
+    if (path.normalize(dirname) === path.normalize(dirPath)) {
```
Comment (Sneezry) [inline src/FileUtility.ts]: When will `dirPath` equal to its parent folder path?

File: src/FileUtility.ts [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+    } else {
+      await FileUtility.mkdirRecursively(type, dirname);
+      await FileUtility.mkdir(type, dirPath);
+    }
+  }
+
+  static async fileExists(type: ScaffoldType, filePath: string): Promise<boolean> {
+    if (type === ScaffoldType.local) {
+      const directoryExists = await sdk.FileSystem.exists(filePath);
+      if (!directoryExists) {5
```
Comment (Sneezry) [inline src/FileUtility.ts]: 5?

File: src/FileUtility.ts [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+        return false;
+      }
+      const isFile = await sdk.FileSystem.isFile(filePath);
+      return isFile;
+    } else {
+      try {
+        return fs.statSync(filePath).isFile();
```
Comment (Sneezry) [inline src/FileUtility.ts]: Do not use sync method

File: src/FileUtility.ts [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+      } catch (e) {
+        return false;
+      }
+    }
+  }
+
+  static async readFile(type: ScaffoldType, filePath: string, encoding: string): Promise<string> {
```
Comment (Sneezry) [inline src/FileUtility.ts]: The general read file interface should not fix the content type to string. If you only read file contains string, better handle at high level.

File: src/constants.ts [hunk 19]
```diff
@@ -57,6 +58,11 @@ export enum OperationType {
...
+  local = 'local',
```
Comment (Sneezry) [inline src/constants.ts]: Better use uppercase

File: src/constants.ts (hunk 10 above)
Comment (dilin-MS) [reply src/constants.ts]: The type is not simply about the environment information.

File: src/Models/AzureComponentConfig.ts (hunk 12 above)
Comment (dilin-MS) [reply src/Models/AzureComponentConfig.ts]: Same as above.

File: src/FileUtility.ts (hunk 13 above)
Comment (dilin-MS) [reply src/FileUtility.ts]: Will fix.

File: src/FileUtility.ts (hunk 14 above)
Comment (dilin-MS) [reply src/FileUtility.ts]: Will fix.

File: src/FileUtility.ts (hunk 15 above)
Comment (dilin-MS) [reply src/FileUtility.ts]: dirPath = `a/b/c/.`
dirname = `a/b/c`

File: src/FileUtility.ts (hunk 16 above)
Comment (dilin-MS) [reply src/FileUtility.ts]: Typo. Will remove.

File: src/FileUtility.ts (hunk 17 above)
Comment (dilin-MS) [reply src/FileUtility.ts]: Fix.

File: src/FileUtility.ts (hunk 18 above)
Comment (dilin-MS) [reply src/FileUtility.ts]: Fixed.

File: src/constants.ts (hunk 19 above)
Comment (dilin-MS) [reply src/constants.ts]: Fixed.

File: src/Models/ScaffoldGenerator.ts [hunk 20]
```diff
@@ -177,54 +75,48 @@ export class ScaffoldGenerator {
...
-      const dockerfileSourcePath = path.join(
-        boardFolderPath, boardId, FileNames.dockerfileName);
-      const dockerfileContent = fs.readFileSync(dockerfileSourcePath, 'utf8');
-      await sdk.FileSystem.writeFile(dockerfileTargetPath, dockerfileContent);
-    } catch (error) {
-      throw new Error(`Create ${FileNames.dockerfileName} failed: ${error.message}`);
+      try {
+        const dockerfileSourcePath = path.join(
+          templateFolderPath, FileNames.dockerfileName);
+        const dockerfileContent = fs.readFileSync(dockerfileSourcePath, 'utf8');
```
Comment (Sneezry) [inline src/Models/ScaffoldGenerator.ts]: Do not use sync method

Comment (dilin-MS) [reply src/Models/ScaffoldGenerator.ts]: The whole refrator will be implemented in other pr.
